free/src/main/scala/cats/free/FreeAp.scala

```diff
@@ -0,0 +1,95 @@
+package cats
+package free
+
+import cats.arrow.NaturalTransformation
+
+object FreeAp {
+
+  /**
+   * Our case classes are the same as those provided by Free with the addition of Ap
+   * which materializes the applicative ap behavior
+   * This is useful when folding the FreeAp instance to a different Monad:
+   * if that targeted monad has an `ap` function that differs from the usual flatMap / map sequence
+   * we will retain that behavior
+   */
+  private final case class Pure[S[_], A](a: A) extends FreeAp[S, A]
+
+  private final case class Suspend[S[_], A](a: S[A]) extends FreeAp[S, A]
+
+  private final case class Gosub[S[_], B, C](c: FreeAp[S, C], f: C => FreeAp[S, B]) extends FreeAp[S, B]
```

Comment:
Sorry for the nitpick, but could we switch the two arguments around to line up with the `Apply.ap` type signature?